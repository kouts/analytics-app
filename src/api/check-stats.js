import { request } from '@/common/request'

export const fetchCheckStats = (checkId, fromTimestamp, toTimestamp) => {
  const dateRangeParams =
    fromTimestamp && toTimestamp
      ? {
          timestamp_gte: fromTimestamp,
          timestamp_lte: toTimestamp
        }
      : {}

  return request({
    url: 'check_stats',
    params: { ...{ checkId }, ...dateRangeParams }
  })
}
