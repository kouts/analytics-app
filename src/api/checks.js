import { request } from '@/common/request'

export const fetchChecks = () => {
  return request({
    url: 'checks'
  })
}
