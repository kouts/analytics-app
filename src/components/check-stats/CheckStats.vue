<template>
  <div v-if="stats.length" class="w-100">
    <div class="row mb-0">
      <div class="col">
        <div class="float-right">
          <button
            type="button"
            :class="['btn btn-primary btn-sm text-white mr-1', chartSelected !== 'response_time' && 'opacity-50']"
            @click="select('response_time')"
          >
            <Typography as="span" bold>Response time</Typography>
          </button>
          <button
            type="button"
            :class="['btn btn-primary btn-sm text-white mr-1', chartSelected !== 'success' && 'opacity-50']"
            @click="select('success')"
          >
            <Typography as="span" bold>Success</Typography>
          </button>
        </div>
      </div>
    </div>
    <div v-if="chartSelected === 'response_time'" class="position-relative">
      <div class="position-absolute d-flex flex-column response-time-filters">
        <button
          v-for="item in responseTimeChartFilters"
          :key="item.name"
          type="button"
          :class="['btn btn-primary btn-sm text-white p-1 mb-1', !item.active && 'opacity-50']"
          @click="item.active = !item.active"
        >
          <Typography as="span" uppercase small>{{ item.name }}</Typography>
        </button>
      </div>
    </div>
    <StatsChart ref="statsChart" :options="chartOptions"></StatsChart>
  </div>
  <div v-else class="d-flex justify-content-center pt-2 pb-4">
    <Typography v-if="!loadingChecks" as="h4">No results found</Typography>
  </div>
</template>

<script>
import Typography from '../library/typography'
import { Chart } from 'highcharts-vue'
import {
  commonChartOptions,
  createResponseTimeSeries,
  createSuccessSeries,
  responseTimeChartOptions,
  successChartOptions
} from './checks-stats.helpers'

export default {
  components: {
    StatsChart: Chart,
    Typography
  },
  props: {
    stats: {
      type: Array,
      default: () => []
    }
  },
  data() {
    return {
      chartOptions: {},
      chartSelected: 'response_time',
      responseTimeChartFilters: [
        {
          name: 'p99',
          active: true
        },
        {
          name: 'p95',
          active: true
        },
        {
          name: 'avg',
          active: true
        }
      ]
    }
  },
  watch: {
    chartSelected: {
      handler(newVal) {
        if (newVal === 'response_time') {
          if (this.chartOptions.title) {
            this.$refs.statsChart.chart.update(this.createResponseTimeChartOptions())
            this.responseTimeChartFilters.map((item) => {
              item.active = true

              return item
            })
          } else {
            this.chartOptions = this.createResponseTimeChartOptions()
          }
        }
        if (newVal === 'success') {
          if (this.chartOptions.title) {
            this.$refs.statsChart.chart.update(this.createSuccessChartOptions())
          } else {
            this.chartOptions = this.createSuccessChartOptions()
          }
        }
      },
      immediate: true
    },
    responseTimeChartFilters: {
      handler(newVal) {
        const showXAxis = newVal.some((item) => item.active)

        newVal.forEach((item) => {
          this.$refs.statsChart.chart.series
            .find((s) => s.name === item.name)
            .update({
              visible: item.active
            })
        })
        this.$refs.statsChart.chart.xAxis[0].update({ visible: showXAxis })
      },
      deep: true
    }
  },
  methods: {
    select(chartToSelect) {
      this.chartSelected = chartToSelect
    },
    createResponseTimeChartOptions() {
      return {
        ...commonChartOptions,
        ...responseTimeChartOptions,
        ...{
          series: [
            createResponseTimeSeries(this.stats, 'p99'),
            createResponseTimeSeries(this.stats, 'p95'),
            createResponseTimeSeries(this.stats, 'avg')
          ]
        }
      }
    },
    createSuccessChartOptions() {
      return {
        ...commonChartOptions,
        ...successChartOptions,
        ...{
          series: [createSuccessSeries(this.stats), { visible: false }, { visible: false }]
        }
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.response-time-filters {
  top: map-get($spacers, 4);
  right: map-get($spacers, 1);
  z-index: 100;
}

.opacity-50 {
  opacity: 0.5;
}
</style>
