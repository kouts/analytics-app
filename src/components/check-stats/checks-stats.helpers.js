import { formatMilliseconds } from '@/common/format-milliseconds'

export const commonChartOptions = {
  title: {
    text: ''
  },
  credits: {
    enabled: false
  },
  legend: {
    enabled: false
  },
  xAxis: [
    {
      visible: true,
      title: false,
      type: 'datetime',
      maxPadding: 0,
      units: [
        ['hour', [0, 3, 6, 12]],
        ['day', [1]]
      ]
    }
  ],
  tooltip: {
    shared: true,
    useHTML: true,
    backgroundColor: '#fff',
    borderWidth: 0,
    crosshairs: true
  },
  plotOptions: {
    area: {
      lineColor: '#2eb3db',
      lineWidth: 1
    },
    series: {
      states: {
        hover: {
          enabled: false
        }
      }
    }
  }
}

export const responseTimeChartOptions = {
  chart: {
    spacingRight: 60,
    type: 'area',
    backgroundColor: 'transparent',
    height: 180
  },
  yAxis: {
    max: null,
    title: false,
    gridLineWidth: 0,
    labels: {
      formatter: function () {
        return formatMilliseconds(this.value)
      }
    }
  }
}

export const successChartOptions = {
  chart: {
    spacingRight: 0,
    type: 'area',
    backgroundColor: 'transparent',
    height: 180
  },
  yAxis: {
    max: 1,
    title: false,
    gridLineWidth: 0,
    labels: {
      formatter: function () {
        return `${Math.round(this.value * 100)}%`
      }
    }
  }
}

export const createResponseTimeSeries = (stats, statAttr) => {
  const data = stats.map((stat) => {
    return { x: stat.timestamp, y: stat[statAttr] }
  })

  return {
    name: statAttr,
    data,
    tooltip: {
      pointFormatter: function () {
        return `<div class="mt-1">${statAttr.toUpperCase()}: <strong>${formatMilliseconds(this.y)}</strong></div>`
      }
    },
    visible: true,
    marker: {
      enabled: false
    },
    fillColor: 'rgba(69, 200, 241, 0.3)'
  }
}

export const createSuccessSeries = (stats) => {
  const data = stats.map((stat) => {
    return { x: stat.timestamp, y: stat.success }
  })

  return {
    data,
    tooltip: {
      pointFormatter: function () {
        return `<div class="mt-1">Success Ratio: <strong>${Math.round(this.y * 100)}%</strong></div>`
      }
    },
    visible: true,
    marker: {
      enabled: false
    },
    fillColor: 'rgba(69, 200, 241, 0.3)'
  }
}
