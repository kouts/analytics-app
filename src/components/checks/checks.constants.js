export const CheckTypes = {
  API: { cssClass: 'badge-check--api' },
  BROWSER: { cssClass: 'badge-check--browser' }
}
