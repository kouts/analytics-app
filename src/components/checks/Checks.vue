<template>
  <table class="table table-checks">
    <thead>
      <tr>
        <th scope="col" class="w-15">TYPE</th>
        <th scope="col" class="w-45">NAME</th>
        <th scope="col" class="w-10 text-center">SUCCESS</th>
        <th scope="col" class="w-10 text-center">AVG</th>
        <th scope="col" class="w-10 text-center">P95</th>
        <th scope="col" class="w-10 text-center">P99</th>
      </tr>
    </thead>
    <tbody v-if="checks.length">
      <template v-for="check in checks">
        <tr :key="check.id">
          <td colspan="6" class="w-100 p-0">
            <table class="table table-borderless mb-0">
              <tr>
                <td class="w-15">
                  <button
                    type="button"
                    class="d-flex align-items-center btn-transparent ml-n2"
                    @click="toggleDetails(check.id, check.expanded)"
                  >
                    <Icon :name="check.expanded ? 'chevron-down' : 'chevron-right'" class="toggle-icon" :width="20" />
                    <span :class="['badge badge-check', CheckTypes[check.type].cssClass]">{{ check.type }}</span>
                    <LoadingSpinner class="ml-1" :size="16" :width="2" :show="loadingStats.includes(check.id)" />
                  </button>
                </td>
                <td class="w-45">{{ check.name }}</td>
                <td class="w-10 text-center">{{ Math.round(check.success * 100) }}%</td>
                <td class="w-10 text-center">{{ formatMilliseconds(check.avg) }}</td>
                <td class="w-10 text-center">{{ formatMilliseconds(check.p95) }}</td>
                <td class="w-10 text-center">{{ formatMilliseconds(check.p99) }}</td>
              </tr>
              <tr v-if="check.expanded" :key="`detail-${check.id}`" class="expanded">
                <td colspan="6" class="w-100 pt-0">
                  <CheckStats :stats="stats[check.id]" />
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </template>
    </tbody>
    <tbody v-else>
      <tr class="no-hover">
        <td colspan="6">
          <div class="d-flex justify-content-center py-4">
            <LoadingSpinner :show="loadingChecks" />
            <Typography v-if="!loadingChecks" as="h3">No results found</Typography>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
import CheckStats from '@/components/check-stats/CheckStats.vue'
import Icon from '@/components/library/icon/Icon'
import LoadingSpinner from '../library/loading-spinner'
import Typography from '../library/typography'
import { CheckTypes } from './checks.constants'
import { fetchCheckStats } from '@/api/check-stats'
import { fetchChecks } from '@/api/checks'
import { formatMilliseconds } from '@/common/format-milliseconds'
import { getLastWeekRange } from '@/common/get-last-week-range'

export default {
  components: {
    Icon,
    CheckStats,
    LoadingSpinner,
    Typography
  },
  data() {
    return {
      checks: [],
      CheckTypes,
      stats: {},
      loadingChecks: false,
      loadingStats: []
    }
  },
  async created() {
    let res = []

    this.loadingChecks = true

    try {
      res = await fetchChecks()
    } catch (error) {
      console.error('Failed to fetch checks')
    }
    res.map((check) => {
      check.expanded = false

      return check
    })
    this.checks = res
    this.loadingChecks = false
  },
  methods: {
    async toggleDetails(checkId, expanded) {
      if (this.loadingStats.includes(checkId)) return

      if (!expanded) {
        // Set the reference date to 24/07/2022
        const referenceDate = new Date(2022, 6, 24)
        const { first, last } = getLastWeekRange(referenceDate)
        const index = this.loadingStats.findIndex((i) => i === checkId)

        try {
          this.loadingStats.push(checkId)
          this.stats[checkId] = await fetchCheckStats(checkId, first, last)
        } catch (error) {
          console.error(`Failed to fetch stats for checkId ${checkId}`)
        }
        this.loadingStats.splice(index, 1)
      }
      const check = this.checks.find((check) => check.id === checkId)

      check.expanded = !check.expanded
    },
    formatMilliseconds
  }
}
</script>

<style lang="scss" scoped>
.w-10 {
  width: 10%;
}

.w-15 {
  width: 15%;
}

.w-45 {
  width: 45%;
}

.toggle-icon {
  color: $gray-500;
}

.table tr.expanded td {
  border-top: 1px solid transparent;
}

.btn-transparent {
  background-color: transparent;
  background-repeat: no-repeat;
  border: none;
  cursor: pointer;
  overflow: hidden;
  outline: none;
}

.badge-check {
  font-size: 0.8rem;
  font-weight: normal;
  color: $white;
  margin-left: map-get($spacers, 1);
  &--api {
    background-color: #353b3a;
  }
  &--browser {
    background-color: #5e6766;
  }
}

.table-checks tbody > tr:not(.no-hover):hover {
  background-color: $gray-100;
}
</style>
