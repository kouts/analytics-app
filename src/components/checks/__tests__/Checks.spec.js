import Checks from '../Checks.vue'
import { fetchCheckStats } from '@/api/check-stats'
import { fetchChecks } from '@/api/checks'
import { mount, shallowMount } from '@vue/test-utils'

const fetchChecksResponse = [
  {
    id: '2b22c08d-b681-43c5-8b23-f49f6168bafe',
    name: 'Check 0',
    type: 'API',
    success: 0.49646666666666667,
    avg: 498.5661333333333,
    p95: 951,
    p99: 990
  }
]

const fetchCheckStatsResponse = [
  {
    checkId: '2b22c08d-b681-43c5-8b23-f49f6168bafe',
    success: 0.3333333333333333,
    avg: 652.3333333333334,
    p95: 781.8,
    p99: 785.16,
    timestamp: 1631019600000
  }
]

jest.mock('@/api/checks', () => ({
  fetchChecks: jest.fn(() => Promise.resolve([]))
}))

jest.mock('@/api/check-stats', () => ({
  fetchCheckStats: jest.fn(() => Promise.resolve([]))
}))

describe('Checks', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should be defined', () => {
    expect(Checks).toBeDefined()
  })

  it('should display a message when there are no checks available', async () => {
    const wrapper = mount(Checks)

    fetchChecks.mockImplementation(() => Promise.resolve([]))
    await wrapper.vm.$nextTick()

    expect(wrapper.find('h3').text()).toBe('No results found')
  })

  it('does not call the fetch stats method upon initialization', async () => {
    fetchChecks.mockImplementation(() => Promise.resolve(fetchChecksResponse))
    fetchCheckStats.mockImplementation(() => Promise.resolve(fetchCheckStatsResponse))
    const wrapper = shallowMount(Checks)

    await wrapper.vm.$nextTick()

    expect(fetchCheckStats).not.toHaveBeenCalled()
  })

  it('calls the fetch stats method when the table row expands', async () => {
    fetchChecks.mockImplementation(() => Promise.resolve(fetchChecksResponse))
    fetchCheckStats.mockImplementation(() => Promise.resolve(fetchCheckStatsResponse))
    const wrapper = shallowMount(Checks)

    await wrapper.vm.$nextTick()
    const item = wrapper.find('button.btn-transparent')

    await item.trigger('click')
    await wrapper.vm.$nextTick()

    expect(fetchCheckStats).toHaveBeenCalled()
  })

  it('renders the stats component when the table row expands', async () => {
    fetchChecks.mockImplementation(() => Promise.resolve(fetchChecksResponse))
    fetchCheckStats.mockImplementation(() => Promise.resolve(fetchCheckStatsResponse))
    const wrapper = shallowMount(Checks)

    await wrapper.vm.$nextTick()
    const item = wrapper.find('button.btn-transparent')

    await item.trigger('click')
    await wrapper.vm.$nextTick()

    const row = item.element.parentNode.parentNode
    const nextRow = row.parentNode.rows[row.rowIndex + 1]

    expect(nextRow.className).toBe('expanded')
    expect(nextRow.querySelector('checkstats-stub')).toBeInstanceOf(HTMLElement)
  })

  describe('snapshots', () => {
    it('should render correctly', async () => {
      fetchChecks.mockImplementation(() => Promise.resolve(fetchChecksResponse))
      const wrapper = mount(Checks)

      await wrapper.vm.$nextTick()

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
