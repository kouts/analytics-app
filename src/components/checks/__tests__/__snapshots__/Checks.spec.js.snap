// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks snapshots should render correctly 1`] = `
<table
  class="table table-checks"
>
  <thead>
    <tr>
      <th
        class="w-15"
        scope="col"
      >
        TYPE
      </th>
       
      <th
        class="w-45"
        scope="col"
      >
        NAME
      </th>
       
      <th
        class="w-10 text-center"
        scope="col"
      >
        SUCCESS
      </th>
       
      <th
        class="w-10 text-center"
        scope="col"
      >
        AVG
      </th>
       
      <th
        class="w-10 text-center"
        scope="col"
      >
        P95
      </th>
       
      <th
        class="w-10 text-center"
        scope="col"
      >
        P99
      </th>
    </tr>
  </thead>
   
  <tbody>
    <tr>
      <td
        class="w-100 p-0"
        colspan="6"
      >
        <table
          class="table table-borderless mb-0"
        >
          <tr>
            <td
              class="w-15"
            >
              <button
                class="d-flex align-items-center btn-transparent ml-n2"
                type="button"
              >
                <div
                  class="icon toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="octicon octicon-chevron-right"
                    height="20"
                    version="1.1"
                    viewBox="0 0 16 16"
                    width="20"
                  >
                    <path
                      d="M6.22 3.22a.75.75 0 011.06 0l4.25 4.25a.75.75 0 010 1.06l-4.25 4.25a.75.75 0 01-1.06-1.06L9.94 8 6.22 4.28a.75.75 0 010-1.06z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
                 
                <span
                  class="badge badge-check badge-check--api"
                >
                  API
                </span>
                 
                <!---->
              </button>
            </td>
             
            <td
              class="w-45"
            >
              Check 0
            </td>
             
            <td
              class="w-10 text-center"
            >
              50%
            </td>
             
            <td
              class="w-10 text-center"
            >
              499ms
            </td>
             
            <td
              class="w-10 text-center"
            >
              951ms
            </td>
             
            <td
              class="w-10 text-center"
            >
              990ms
            </td>
          </tr>
           
          <!---->
        </table>
      </td>
    </tr>
  </tbody>
</table>
`;
